Add critical path guide for CI Visibility (#26408)
DataDog/documentation: content/en/continuous_integration/guides/identify_highest_impact_jobs_with_critical_path.md

---
title: Identify CI Jobs on the Critical Path to reduce the Pipeline Duration
description: Learn how to identify those CI Jobs that are on the critical path to improve the duration of your CI Pipelines.
further_reading:
  - link: "/continuous_integration/search/#pipeline-details-and-executions"
    tag: "Documentation"
    text: "Learn how to search and manage your pipeline executions"
  - link: "continuous_integration/search/#highlight-critical-path"
    tag: "Documentation"
    text: "Highlight critical path in your Pipeline Execution"
---

## Overview

This guide explains how to identify the CI jobs that are on the critical path, with the goal of determining which CI jobs to focus on to reduce the overall duration of the CI pipelines.

### Understanding the Critical Path in a CI Pipeline

> The critical path in a CI Pipeline is the longest sequence of CI Jobs that determines the total duration of that pipeline. In other words, it is the path through the DAG of CI Jobs that requires the most time to complete. You can only reduce the total duration of a CI Pipeline by reducing the duration of CI Jobs that are on the critical path.

Reviewer: ```suggestion
The critical path of a CI Pipeline execution is the longest sequence of CI Jobs that determines the total duration of that pipeline execution. In other words, it is the path through the dependency graph of CI Jobs that took the most time to complete. You can only reduce the total duration of a CI Pipeline execution by reducing the duration of CI Jobs that are on the critical path.
```

{{< img src="continuous_integration/critical_path_highlight_pipeline.png" alt="Highlight of jobs on the critical path in a pipeline execution." width="90%">}}

> However, CI jobs are typically executed in parallel with other jobs, meaning that the actual reduction in pipeline time is determined by the **exclusive time** a CI job spends on the CI runner during its execution.

Reviewer: ```suggestion
Looking at the job duration may not be enough. CI jobs are typically executed in parallel with other jobs, meaning that the reduction of the pipeline execution time is determined by reducing the **exclusive time** of the CI job.
```

> The exclusive time that a job is on the critical path represents the amount of time the CI runner has spent executing a specific job exclusively, excluding the execution time of other jobs that were running in parallel.

Reviewer: ```suggestion
The job **exclusive time** on the critical path represents the amount of time the CI runner has spent executing a specific job exclusively, excluding the execution time of other jobs that were running in parallel.
```

{{< img src="continuous_integration/critical_path_highlight_pipeline_exclusive_time.png" alt="Highlight exclusive time of the jobs on the critical path in a pipeline execution." width="90%">}}

> If CI job A is on the critical path with a duration of 100ms and runs in parallel with CI job B, which takes 80ms, the exclusive time for job A on the critical path would be 20ms. This means the actual improvement in pipeline duration reduction would be 20ms.

Reviewer: ```suggestion
If CI job A is on the critical path with a duration of 100ms and runs in parallel with CI job B, which duration is 80ms, then job A exclusive time on the critical path is 20ms. Even if job A duration was reduced to 1ms, the pipeline duration would be reduced by only 20ms.
```

> ### Compatibility

Reviewer: ```suggestion
### Supported CI Providers
```

> If CI job A is on the critical path with a duration of 100ms and runs in parallel with CI job B, which has a duration of 80ms, the exclusive time of CI job A on the critical path is 20ms. This means that reducing the duration of the CI Job A by more than 20ms would still only decrease the overall pipeline duration by 20ms.

Reviewer 2: NIT: I'd use `J` and `J1` here, in a code-notation

> If CI job A is on the critical path with a duration of 100ms and runs in parallel with CI job B, which has a duration of 80ms, the exclusive time of CI job A on the critical path is 20ms. This means that reducing the duration of the CI Job A by more than 20ms would still only decrease the overall pipeline duration by 20ms.

Reviewer 2: This is true only if the entire job `B` (`J1`) execution is done in parallel with Job `A` (`J`), right? So for example if they start at the same time

Author: It's not strictly needed to start at the same time, but yeah, both jobs need to run in parallel.

> <div class="alert alert-info">Are you interested in critical path and your CI provider is not supported yet? Fill out <a href="https://forms.gle/zDgiDSGQxA1HgjXh6" target="_blank">this form</a>.</div>

Reviewer 2: I would put this after the sentence in which we state for which providers this is enabled, otherwise they don't know 😬

Filtering and Exclusive Time computation of CI Jobs on the Critical Path is available for the following CI Providers:
* [GitLab][3]

> You can still [highlight which CI Jobs are on the critical path][4] using the Pipeline Execution detail view across all CI providers.

Reviewer: Can you please add a link to this guide the other way around, in https://docs.datadoghq.com/continuous_integration/search/#highlight-critical-path ?

Author: Added

- Open [civisibility-critical-path-gitlab-dashboard.json][1] dashboard template and copy the content in the clipboard.
- Create a [New Dashboard][2] in Datadog.
- Paste the copied content in the new dashboard.
- Save the dashboard.

{{< img src="continuous_integration/critical_path_dashboard.png" alt="Critical path dashboard for CI Visibility" width="90%">}}

#### Terminology

| Column                                | Description                                                                                                                                                      |
|---------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Total Exclusive Time On Critical Path | Sum all exclusive time of the job. It gives an approximation of the potential saved time the involved pipelines might have.                                      |
| Avg Exclusive Time On Critical Path   | Avg exclusive time of a particular job on the critical path. This measures the potential reduction of a pipeline duration if the job reduces its exclusive time. |

> | Rate On Critical Path                 | Measures the porcentage that a job is on the critical path compared when that job is not.                                                                        |

Reviewer: ```suggestion
| Rate On Critical Path                 | Measures how often a job is on the critical path.                                                                        |
```

##### Example

In the previous image, we can observe that a CI job called `metrics` is a potential candidate for improvement, as its total exclusive time is the highest. The average exclusive time is around 21 minutes, meaning there is room for improvement of up to 21 minutes for this CI job. 

> Since we know this CI job is on the critical path, any improvements made to it will be reflected in the corresponding CI pipelines, reducing a similar amount of time.

Reviewer: ```suggestion
Since we know this CI job is on the critical path 43.5% of the time, we can hope to reduce the average pipeline duration by 21 minutes for 43.5% of the pipeline executions.
```

> Since we know this CI job is on the critical path 43.5% of the time, we can hope to reduce the average pipeline duration by 21 minutes for 43.5% of the pipeline executions.

Reviewer 2: ```suggestion
Since we know this CI job is on the critical path 43.5% of the times, we could potentially reduce the average pipeline duration by up to 21 minutes for 43.5% of the pipeline executions.
```

> {{< img src="continuous_integration/critical_path_dashboard_outlier_job_hightlighted.png" alt="Potencial CI Job candidate to improve the exclusive time." width="90%">}}

Reviewer 2: ```suggestion
{{< img src="continuous_integration/critical_path_dashboard_outlier_job_highlighted.png" alt="Potential CI Job candidate to improve the exclusive time." width="90%">}}
```
